src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,1287 @@
+import math
+import os
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Iterable, List, Mapping, Optional, Sequence, Tuple, Union
+
+import fiona
+import laspy
+import rasterio
+import shapely
+from rtree import index
+
+import tiledb
+from tiledb.cloud.utilities import Profiler
+from tiledb.cloud.utilities import as_batch
+from tiledb.cloud.utilities import batch
+from tiledb.cloud.utilities import create_log_array
+from tiledb.cloud.utilities import find
+from tiledb.cloud.utilities import get_logger_wrapper
+from tiledb.cloud.utilities import max_memory_usage
+from tiledb.cloud.utilities import run_dag
+
+fiona.drvsupport.supported_drivers["TileDB"] = "arw"
+fiona.vfs.SCHEMES["tiledb"] = "tiledb"
+
+DEFAULT_RESOURCES = {"cpu": "2", "memory": "2Gi"}
+DEFAULT_IMG_NAME = "3.9-geo"
+DEFAULT_DAG_NAME = "geo-ingestion"
+
+# Default values for ingestion parameters
+RASTER_TILE_SIZE = 1024
+POINT_CLOUD_CHUNK_SIZE = 1_000_000
+GEOMETRY_CHUNK_SIZE = 100_000
+MAX_WORKERS = 40
+BATCH_SIZE = 10
+
+
+class DatasetType(Enum):
+    POINTCLOUD = 1
+    RASTER = 2
+    GEOMETRY = 3
+
+
+@dataclass
+class BoundingBox:
+    minx: float
+    miny: float
+    maxx: float
+    maxy: float
+    minz: Optional[float] = None
+    maxz: Optional[float] = None
+
+    @property
+    def bounds(self):
+        if self.minz:
+            return (self.minx, self.miny, self.minz, self.maxx, self.maxy, self.maxz)
+        else:
+            return (self.minx, self.miny, self.maxx, self.maxy)
+
+
+@dataclass
+class GeoBlockMetadata:
+    ranges: Tuple[Tuple[int, int], Tuple[int, int]]
+    files: Tuple[os.PathLike, ...]
```

Comment:
using `attrs` will also make this a bit nicer, since you can use a converter here:

```python
files: Tuple[os.PathLike, ...] = attrs.field(converter=tuple)
```

This will allow you to pass any Iterable in and it will be converted to a tuple.

Reply:
👍